apps/api/src/database/repositories/competition-repository.ts

```diff
@@ -725,6 +727,69 @@ export async function getAgentPortfolioSnapshots(
   }
 }
 
+/**
+ * Get the newest and oldest portfolio snapshots for an agent in a competition
+ * @param competitionId Competition ID
+ * @param agentId Agent ID
+ * @returns Object with newest and oldest snapshots, or null if no snapshots found
+ */
+export async function getBoundedSnapshots(
+  competitionId: string,
+  agentId: string,
+) {
+  try {
+    // Create subqueries for newest and oldest snapshots
+    const newestQuery = db
+      .select()
+      .from(portfolioSnapshots)
+      .where(
+        and(
+          eq(portfolioSnapshots.competitionId, competitionId),
+          eq(portfolioSnapshots.agentId, agentId),
+        ),
+      )
+      .orderBy(desc(portfolioSnapshots.timestamp))
+      .limit(1);
+
+    const oldestQuery = db
+      .select()
+      .from(portfolioSnapshots)
+      .where(
+        and(
+          eq(portfolioSnapshots.competitionId, competitionId),
+          eq(portfolioSnapshots.agentId, agentId),
+        ),
+      )
+      .orderBy(asc(portfolioSnapshots.timestamp))
+      .limit(1);
+
+    // Union the queries and execute
+    const results = await unionAll(newestQuery, oldestQuery);
+
+    if (results.length === 0) {
+      return null;
+    }
+
+    // Sort results by timestamp desc to identify newest vs oldest
+    const sortedResults = results.sort(
+      // timestamp has default value of now() + is not nullable, so safe to use `!`
```

Comment:
oooh, actually, the `portfolio_snapshots` table technically _doesn't_ explicitly set the `timestamp` column as non-nullable...it [does this](https://github.com/recallnet/js-recall/blob/main/apps/api/src/database/schema/trading/defs.ts#L180):
```ts
timestamp({ withTimezone: true }).defaultNow()
```

so i think we should just fix the `portfolio_snapshots` table:
1. update the `timestamp` col definition to also include `notNull()`
2. then, we can remove the bang operator here